retry/exp.go

```diff
@@ -0,0 +1,57 @@
+package retry
+
+import (
+	"time"
+
+	"github.com/juju/utils/clock"
+)
+
+// Exponential represents an exponential backoff retry strategy.
+// To limit the number of attempts or their overall duration, wrap
+// this in LimitCount or LimitDuration.
+type Exponential struct {
+	// Initial holds the initial delay.
+	Initial time.Duration
+	// Factor holds the factor that the delay time will be multiplied
+	// by on each iteration.
+	Factor float64
+	// MaxDelay holds the maximum delay between the start
+	// of attempts. If this is zero, there is no maximum delay.
+	MaxDelay time.Duration
+}
+
+type exponentialTimer struct {
+	strategy Exponential
+	start    time.Time
+	end      time.Time
+	delay    time.Duration
+}
+
+// Start is short for Start(r, clk, stop).
+func (r Exponential) Start(clk clock.Clock, stop <-chan struct{}) *Attempt {
```

Comment:
ew. If we can avoid the default case being (nil, nil), that would be really nice. Why not make the clock part of the struct? Surely no one is going to need to change out the clock when reusing an attempt strategy.

Then you could just have Start and StartWithCancel or something.